### ISSUE TITLE
Docker timeout on startup

### ISSUE THREAD

**Issue by fabiogusmao** [docker/for-win#989]
I've seen many posts with this same issue on this repository so I decided to open a new one just with the answer so that it is more easily detectable by the Docker for Windows team if they actually read these issues (let's hope so!)

The steps to reproduce are very well described in #941. 

# TL;DR

Docker takes forever to start and after many minutes it presents a crash report with the message:

> Docker command did not complete within the allowed timeout

Solved by adding **Docker for Windows.exe** as an allowed program in Windows Firewall, as suggested by [this solution](https://forums.docker.com/t/failed-to-contact-the-database-on-pipe-dockerdatabase/17345/14)

# Explanation

Windows Firewall blocks **Docker for Windows.exe** from opening the docker pipe if it is running for the first time (?) and the program prints the following message every 100ms on the log file:

> Failed to contact the database on \\.\pipe\dockerDataBase

In my computer there were 2 programs registered in Windows Firewall:

- DockerProxy
- DockerSmbMount

Adding **Docker for Windows.exe** as a 3rd program solves the problem.

So please make this change to the Docker for Windows installer and the world will be in peace once again :)

**Comment by jasonbivins** [comment]
Hi @fabiogusmao   Thanks for posting this!  I'll add an enhancement request, and write up a quick KB 

Would you mind posting a diagnostic ID so I can get the exact error you were seeing?  Right click the whale icon - Diagnose and Feedback

**Comment by bzuu** [comment]
Hi @jasonbivins Thanks for posting your solution. It helped me resolve the issue.

Perhaps good to add an additional note however that the Windows Firewall rule should allow all protocols not just TCP/UDP (which is the default if you create the rule using the **Allow an app or feature through Windows Firewall** link). When you create the rule in the **Advanced Settings** window it will by default allow all protocols.

**Comment by jasonbivins** [comment]
@bzuu 
Something like this?

``` 

New-NetFirewallRule -DisplayName Docker for Windows `
    -Program “C:\Program Files\Docker\Docker\Docker for Windows.exe” `
    -Action Allow `
    -Profile Domain, Private `
    -Description “Docker” `
    -Protocol Any `
    -Direction Inbound ```

**Comment by bzuu** [comment]
@jasonbivins
Although your suggested approach for adding the firewall matches what I sais, unfortunately it seems the firewall rule did not resolve the problem for me after all.

Apparently the behaviour is different whether I just restart Windows or if I give it a full shutdown (i.e. properly power off). In the former case Docker starts just fine, in the latter it gives me this error:
```
Docker command did not complete within the allowed timeout. Output: 
   at Docker.Backend.MobyCommand.RunDocker(String arguments, IDictionary`2 envVariables, String stdin, Int32 timeoutMs, String cidFile)
   at Docker.Backend.MobyCommand.Run(String commandLine, IDictionary`2 envVariables, Boolean load, String stdin, Int32 timeoutMs)
   at Docker.Backend.SambaShare.MountInMoby(String mount, String drive, Credential cred, CredentialInfo credInfo, String options, String ip)
   at Docker.Backend.SambaShare.MountInMoby(String drive, Credential credential, Settings settings)
   at Docker.Backend.SambaShare.Mount(String drive, Credential credential, Settings settings)
   at Docker.Backend.BackendServer.<Run>b__5_11(Object[] args)
   at Docker.Core.Pipe.NamedPipeServer.RunAction(String action, Object[] parameters)
```
When I added the firewall rules yesterday I restarted Windows (not a full shutdown) and wrongly concluded the issue was resolved.
I uploaded a diagnostic with ID 8178649A-8F95-4DC1-A293-2BE51F0D1D09/2017-08-11_09-25-48 so you can further investigate.

**Comment by bzuu** [comment]
@jasonbivins Yet another update... after completely reinstalling Docker the issue is gone. So perhaps this was in part related to a previous update going wrong?

**Comment by fabiogusmao** [comment]
@jasonbivins I did all my tests the other day with  **Start Docker when you log in** unchecked and later enabled it but yesterday I noticed the timeout message appears again as @bzuu mentioned so for me the behavior is as follows:

- **Start Docker when you log in** enabled: I get the message **Docker command did not complete within the allowed timeout**.
- **Start Docker when you log in** disabled: works fine.

I'll try reinstalling Docker as @bzuu mentioned and post results / diagnostics here.

**Comment by fabiogusmao** [comment]
OK @jasonbivins , sit back and relax because here comes a story with lots of Diagnostic IDs (lol)

## TL;DR

Problems indeed happens when machine is powered on; starts OK when restarting Windows from the Start Menu

## Long story

I ocmpletely reinstalled Docker as @bzuu did (**89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_10-24-47**) and it behaved OK.

I shared the D:\ drive where my source code is and then restarted and it worked OK (**89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_10-28-57**).

Then I ran `docker-compose up` to create the containers for my main project and it was still OK (**89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_10-42-16**)

I turned off the computer and waited a few seconds before turning it on again and this time Docker would take forever to start. I did not wait Docker to give up but instead captured the 
[log file](https://github.com/docker/for-win/files/1218061/shutdown.txt).

Another restart and it works fine again (**89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_10-52-52**).

Finally, I turned it off once again but this time waited for Docker to give up while writing this message and got an ID **89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_11-02-17**.

Finally [2] I clicked **Restart Docker** on the Settings window and it was alrigght with ID **89A1A798-449A-44A8-930B-FD25E850346C/2017-08-11_11-09-40**.

## Conclusion

This problem may or may not be related to firewall exception, may go away if you restart the computer without turning it off and most likely depends on some other service or resource that Windows fails to provide when starting up cold.

**Comment by jasonbivins** [comment]
Hi @fabiogusmao   Thank you for the thorough write up!  I dug through the logs a bit, and it does seem to be a Startup issue not a firewall issue.  It does appear that the Docker for Windows service is starting up out of order when it starts with Windows.  Here's the error it throws when it times out - which is consisten with the Docker service not being connected to the network:

```[11:09:22.602][ApiProxy       ][Info   ] error reading response from Docker:  HvSocket has already been closed
[11:09:22.610][DockerDaemonChecker][Error  ] Docker daemon is not running
[11:09:22.745][DataKit        ][Error  ] com.docker.db.exe: Hvsock.read raised Unix.Unix_error(Unix.ECONNABORTED, "read", ""): returning `Eof```

**Comment by jasonbivins** [comment]
@bzuu   You're getting the same error, see below.  The best workaround for the time being seems to be disabling the auto-start with Windows, and starting Docker manually.

```[09:19:44.558][ApiProxy       ][Info   ] error reading response from Docker:  HvSocket has already been closed
[09:19:44.566][DockerDaemonChecker][Error  ] Docker daemon is not running
[09:19:44.709][DataKit        ][Error  ] com.docker.db.exe: Hvsock.read raised Unix.Unix_error(Unix.ECONNABORTED, "read", ""): returning `Eof

[09:19:44.715][Moby           ][Info   ]  [ ok ]
```

**Comment by jeffbarnes769** [comment]
bcdedit /set hypervisorlaunchtype auto fixes the docker on startup problem as well

Thanks to user: https://stackoverflow.com/users/2364438/tvs for pointing this out in another thread

**Comment by fabiogusmao** [comment]
Issue still happens on Docker version 17.09.0-ce-rc2, build 363a3e7

Diagnostic ID:
FD25E850346C/2017-09-21_10-50-43

The way this happens to me is stil lthe same:

1. Start up Windows
2. Do some work (view email, open Visual Studio etc)
3. Start Docker from shortcut on Desktop
4. Wait wait wait
5. View log (cannot open database)
6. See error window
7. Restart Docker
8. Everything is fine

The bad part is that the initialization process has to retry connecting to the database many times before it gives up and it is not possible to restart docker while this is going on. Takes around 5 minutes.

This issue does not happen in the computer at the office but used to happen on this one before I reinstalled Windows 1 month ago when adding an extra SSD. The other computer is a 2nd gen Core i5 and this one is 6th gen Core i5. Does that impact in anything? Both are Windows 10 Pro x64.

**Comment by fabiogusmao** [comment]
The first time Docker starts after an update it runs normally (with or without requiring a system reboot)

**Comment by fabiogusmao** [comment]
This is OK now but now we have #1330 
Closing this one since it no longer happens.

**Comment by odulkanberoglu** [comment]
Still having the issue with Docker Version 17.11.0-ce-win40 (14328)

Diagnostic ID:
379E0C00-2B18-4FCA-BDF3-1AFAE543F5D0/2017-11-23_22-51-32

After restarting Docker, runs fine.

**Comment by fabiogusmao** [comment]
I too noticed it continued happening, although not every time as before.

**Comment by abjbhat** [comment]
Still not solved. 

If I add the `graph` option in `daemon.json`, Docker crashes while restarting. I had to reset Docker for Windows back. 

I'm using Version 17.12.0-ce-win47 (15139)

**Comment by abjbhat** [comment]
This is no longer a problem. When I reinstalled Docker for Windows this time, the installer correctly picked up the HyperV default directory (which I have changed) and placed the MobyLinux VHD in that location. No extra steps were needed.
